### DISCUSSION TITLE
Welcome to hlg-tools Discussions!

### DISCUSSION THREAD

**Discussion by wswartzendruber** [wswartzendruber/hlg-tools#1] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by TomHttp451** [comment]
Eagerly waiting for executables ! :)

**Reply by wswartzendruber** [reply]
They're right here:

[https://github.com/wswartzendruber/hlg-tools/releases](url)

You have to expend the "Assets" link in order to see them. I only have builds for Linux and Windows so far.

**Comment by TomHttp451** [comment]
Many thanks, I have downloaded the Linux version.
This is my first use of github and I'm still reading the newbie doc.

**Comment by ChrisCrashX** [comment]
Can you explain how to use hlgprev.ps1 on windows 10 because I cant get it to work no matter what I type ;~;

**Reply by wswartzendruber** [reply]
Try `.\hlgprev.ps1` while being in the same directory as the script.

**Reply by ChrisCrashX** [reply]
![powershell_wjJbchDKMt](https://user-images.githubusercontent.com/30630017/163100228-773687a2-8c17-4bad-ba03-22a552125eae.png)

**Reply by wswartzendruber** [reply]
# Procedure

While `hlg-tools` supports both Linux and Windows, this guide will assume the use of Linux. If
using Windows:

1. Add `.exe` to executable file names.
2. Change any `.sh` extension to `.ps1`.

Note that if `pqstat` needs to be used, then that command should be run from Command Prompt
rather than PowerShell due to the pipe operator.

## Prerequisites

Now let's walk through converting a 4K UltraHD Blu-ray to HLG. For this scenario, we'll be using
`hlg-tools` along with `ffmpeg` and VLC. In particular, we'll be assuming that the following
binaries and scripts are in the active `PATH`:

- `pq2hlg` (part of the `hlg-tools` package)
- `pqstat` (part of the `hlg-tools` package)
- `hlgprev.sh` (part of the `hlg-tools` package)
- `sdrprev.sh` (part of the `hlg-tools` package)
- `ffmpeg` (provided by distribution or third party)

**Reply by isaacolsen94** [reply]
How do you go about installing/using this tool in Ubuntu? When I try I get this message: 
![image](https://github.com/wswartzendruber/hlg-tools/assets/52682311/6b658716-ce3a-42d3-9b3d-ac4e0eaeb0d0)

**Reply by wswartzendruber** [reply]
I'm not sure how you got to have a file called `hlg2pq.sh`, but that's not part of the archive you downloaded. Go ahead and remove everything there except for the `.tar.xz` file and re-extract that.

Then, from the directory with the extracted files, execute:

`./hlg2pq hlg2pq.cube`

This will leave you with a file called `hlg2pq.cube`. That file can be fed into a conversion program that accepts 3D LUTs to convert your HLG to PQ.

**Reply by isaacolsen94** [reply]
Thank you! I was able to get the hlg2pq.cube file. What steps would I do next and which software should I use to convert? Also how to a run it on the file I want to covert from hlg to pq? Thank you again for your help!

**Reply by wswartzendruber** [reply]
I'm getting ready to release version 2.0.0 in the coming days. I'll have some documentation on hlg2pq.

By the way, what are you trying to convert? And what device are you trying to view it on?

**Reply by isaacolsen94** [reply]
Oh awesome! That would be great! I'm trying to convert a 4k hlg auto race so I can watch it from my Plex server on my Nvidia shield. Plex doesn't seem to support HLG so I'm trying to convert to a format Plex can play natively

**Reply by isaacolsen94** [reply]
So I was able to get it to run, but realized after the fact that I was following the steps for the normal PQ to HLG instead of HLG to PQ. I got it to work by running the following command: 
`ffmpeg -i video.mkv -vf scale=3840:2160 -pix_fmt yuv420p10le -f rawvideo - | ~/Downloads/hlg-tools/pqstat -w 3840 -h 2160 `

I had to change the `rgb481e` to `yuv420p10le`, but then it ran. I only got a `MaxCLL` value of `1` which is when I realized I was following the wrong process :P Do I do the steps in reverse? I can wait until 2.0 but I thought I'd let you know where I got to with our discussion so far.

**Reply by wswartzendruber** [reply]
Look for examples of encoding HDR10 with FFmpeg. You need to tell the encoder that the transfer function is ST.2084. Put MaxCLL at 1,000 and MaxFALL at something like 400.

**Comment by isaacolsen94** [comment]
> I'm getting ready to release version 2.0.0 in the coming days. I'll have some documentation on hlg2pq.
> 
> By the way, what are you trying to convert? And what device are you trying to view it on?

Hey! Hope you had a great holiday. Any news on the help documents for hlg2pq? I finally got ffmpeg/handbrake to convert hlg to pq, but they turned out dark so looking forward to your solution

**Reply by wswartzendruber** [reply]
I was working on 2.0.0's documentation just a couple of hours ago.

**Reply by wswartzendruber** [reply]
This should be enough to get you started:

https://github.com/wswartzendruber/hlg-tools/wiki/hlg2pq

**Reply by isaacolsen94** [reply]
Thank you so much! I'll be diving into this tonight. Thanks for the wiki!

**Reply by wswartzendruber** [reply]
Bah. Play with the wiki page, sure, but I'm seeing some problems with color variance. I need to spend more time testing hlg2pq (which isn't a main area of focus).

**Reply by isaacolsen94** [reply]
I don't know anything, but I'll include the ffmpeg command I found somewhere that seemed to work but I couldn't verify the difference between the two files since my display won't do HLG (hence the whole effort). But they seem to come out well. Sometimes the color seems a bit bright but again can't compare. 

Command: 
`ffmpeg -hwaccel cuda -i input.mkv -vf zscale=transfer=smpte2084 -profile:v main10 -tier high -pix_fmt p010le -c:v hevc_nvenc -y -hide_banner -cbr true -b:v XXM -maxrate XXM -minrate XXM -bufsize XM -b_ref_mode 1 -c:a copy "SXXEXX - output (2160p50 HDR).mkv"`

The important bit I believe is the `-vf zscale=transfer=smpte2084` and `-pix_fmt p010le` 

But again I just found this and it seemed to work. I don't actually know how so don't know if it will help you out

**Reply by wswartzendruber** [reply]
`zscale` resizes the image. Try the example in the wiki entry I just wrote. I'm not seeing anything in there that actually converts from PQ to HLG. `-vf zscale=transfer=smpte2084` is going to take PQ into account when resizing the image, but it's not a PQ video.

**Reply by isaacolsen94** [reply]
All I know is that if I run that command media info reads the input.mkv as HLG and the output.mkv as PQ. And my tv registers it correctly and plays in HDR instead of SDR like it was before running. Maybe the transfer=smpte2084 part of zscale does it? I don't know, again I only found this after looking for ever and combining a bunch of different forum posts. 
 
 I plan on running the command and the one from the wiki overnight so I can compare them tomorrow

**Reply by wswartzendruber** [reply]
Yes, because you're setting the bitstream flags to signal a PQ video, but you're not actually converting the color.

EDIT: I've added unit tests that confirm correct HLG<->PQ conversion. I'm not sure why I saw color variance earlier.
